### ISSUE TITLE
Directory listing is still super slow even after all the improvements made

### ISSUE THREAD

**Issue by gyurix** [javi11/usenet-drive#153]
Currently the software is only usable with rclone read-only mounts if there are a very limited amount of nzbs in the same folder and there aren't many folders. 5 files + 30 folders in a single folder and having a total of max 2000 nzb files is what the application is able to handle with rclone mounts in a somewhat stable way.

The goal would be to have no significant performance impact of have 100k file / folder combination in the same folder and a total of 10s of millions of nzb files. The first step of achieving this would be replacing the Webdav with SFTP.

**Comment by gyurix** [comment]
@javi11 We had a discussion about this software with @zozidalom and we would like to motivate you with a bigger donation if you would be able to properly sort out the following git issues:
https://github.com/javi11/usenet-drive/issues/153 (this one)
https://github.com/javi11/usenet-drive/issues/152
https://github.com/javi11/usenet-drive/issues/151
https://github.com/javi11/usenet-drive/issues/148
https://github.com/javi11/usenet-drive/issues/147
https://github.com/javi11/usenet-drive/issues/127

All 6 of these tickets are kinda mandatory for the proper working of the application. Our goal is to upload multiple petabytes of data with it and then be able to mount that data for easy access so we would prefer to see the development of the application going towards that direction instead of wasting too much time on useless features such as web interface improvements and stuff like that.

**Comment by gyurix** [comment]
@MythodeaLoL What exact rclone flags are you using for the mount? How many files do you have per folder? How many folders do you have inside a folder? Are you able to immediately access the files after the mount is made or you need to wait multiple minutes for them to cache up in the rclone directory cache in order to get fast directory listing speed? For me the situation is so bad, that not only the mounts are broken, but also directory listing in general.

I can't rclone copy from any folder including a bunch of nzb files, because it will never start loading. I need to do a find * -type f -exec command on the nzbs folder or any of it's subfolders I want to download and then use the rclone moveto command to do copy operation on a single file instead of the folder. And then I need to make all of that open on multiple screens in order to max out the download speed, because somehow downloading a single file at a time is very slow, but downloading multiple files at a time is fast. It makes no sense that it is like that, but that's what's happening.

`find /root/UsenetDrive/nzbs/subfolder/* -type f -exec bash -c 'while [ $(screen -ls | grep -c "^\s*[0-9]\+\.\w\+") -ge 30 ]; do sleep 5; done ; screen -dmS rcx rclone moveto -v --stats 10s "un:subfolder${0:29:}" "/root/targetdir${0:29:-4}" ' "{}" \;

I need to use this command or something similar to be able to download the files with the optimal speed. In order to resolve this ticket I would like to see copying a folder from the mount have the same speed as this super complicated command.

**Comment by javi11** [comment]
I will test with 2000k of nzb and see what I can do. The problem here is that any filesystem will perform a download of some parts for every file on a directory when you access it and that can cause this slowness in case of a largue amount of files.

Regarding the slow download of 1 file, @gyurix, could you please do a rclone copy of one of the files and show here which are the speeds you reached? Also the number of download workers you used and connections.

As an example I have 100 connections of 2 different providers using 20 download workers and I can achieve speeds of 60 mb/s for one download

**Comment by gyurix** [comment]
@javi11 The actual speed heavily depends on the provider as well, but with tweaknews 30 connections, 30 download workers I get around 15-30 MB/s when I download 1 file after restarting UsenetDrive, after a few hours it goes down to around 4-5MB/s. When I have the exact same config but do 30 downloads in parallel I am able to max out my gigabit download speed. With other providers I get even slower speeds, sometimes speeds around 5-6MB/s when the program starts and 1-2MB/s after a few hour of it running, but even on those providers without any config change increasing the number of files copied in parallel drastically bumps up the speed, sometimes even in a completely linear way, i.e. 10 files copied in parallel = 10x the speed of 1 file copied.

**Comment by gyurix** [comment]
Also @javi11 based on what I read online I do believe this directory listing is an issue of the WEBDav servers themselves too to some degree, so I would definitely propose changing WEBDav to SFTP or making the 2 coexist / config toggleable.

**Comment by javi11** [comment]
Yes, I was thinking on that too I will make this configurable, thanks for the idea.
